### PR TITLE
cleanup: build with -Wconversion

### PR DESCRIPTION
**Pull request by coryan**
Enable `-Wconversion -Wno-sign-conversion` on the CMake builds. We
discussed this as maybe introducing too many casts, but also maybe
helping us catch stuff.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4907)

### CODE REVIEW

File: google/cloud/bigquery/internal/streaming_read_result_source.cc [hunk 1]
```diff
@@ -55,7 +55,8 @@ StatusOr<absl::optional<Row>> StreamingReadResultSource::NextRow() {
   fraction_consumed_ =
       progress.at_response_start() +
       (progress.at_response_end() - progress.at_response_start()) *
-          offset_in_curr_response_ * 1.0 / curr_->row_count();
+          static_cast<double>(offset_in_curr_response_) * 1.0 /
```
Comment (devbww) [inline google/cloud/bigquery/internal/streaming_read_result_source.cc:58]: `s/ \* 1\.0//`?

File: google/cloud/spanner/numeric.h [hunk 2]
```diff
@@ -214,11 +214,11 @@ StatusOr<T> ToInteger(Numeric const& n, int exponent = 0) {
   constexpr auto kMax = (std::numeric_limits<T>::max)();
   enum { kIntPart, kFracPart } state = kIntPart;
   for (auto const ch : rep) {
-    auto dp = std::strchr(kDigits, ch);
+    auto const* dp = std::strchr(kDigits, ch);
     if (state == kFracPart) {
       if (dp - kDigits >= 5) {  // dp != nullptr
         if (v == kMax) return internal::DataLoss(rep);
-        v += 1;
+        v = static_cast<T>(v + 1);
```
Comment (devbww) [inline google/cloud/spanner/numeric.h:221]: Observation: The extra annoying thing about these ones is that the previous statement always ensures that the result of the rhs expression is indeed representable in the type of the lhs.

Comment (coryan) [reply google/cloud/spanner/numeric.h:221]: Ack.

File: google/cloud/bigquery/internal/streaming_read_result_source.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/bigquery/internal/streaming_read_result_source.cc]: Thanks, fixed.

File: google/cloud/spanner/numeric.h (hunk 2 above)
Comment (devjgm) [reply google/cloud/spanner/numeric.h:221]: No change necessary, but for future reference, would `v += T{1}` work, or would `-Wconversion` complain about that too?

Comment (coryan) [reply google/cloud/spanner/numeric.h:221]: I just tested again and `v += T{1}` and `v += static_cast<T>(1)` both produce warnings:

```
../google/cloud/spanner/numeric_test.cc:389:3:   required from here
../google/cloud/spanner/numeric.h:221:11: error: conversion from ‘int’ to ‘unsigned char’ may change value [-Werror=conversion]
  221 |         v += T{1}; // static_cast<T>(v + 1);
      |         ~~^~~~~~~
```

File: google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc [hunk 3]
```diff
@@ -163,9 +163,9 @@ int main(int argc, char* argv[]) {
     auto const r =
         RunOneIteration(generator, *options, bucket_name, object_names);
     std::cout << r.bytes << ',' << r.elapsed.count() << std::endl;
-    auto const mi_b = r.bytes / gcs_bm::kMiB;
-    auto const mi_bs =
-        mi_b * (1.0 * decltype(r.elapsed)::period::den) / r.elapsed.count();
+    auto const mi_b = static_cast<double>(r.bytes) / gcs_bm::kMiB;
+    auto const mi_bs = mi_b * (1.0 * decltype(r.elapsed)::period::den) /
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/benchmarks/storage_shard_throughput_benchmark.cc]: Fixed.

**Pull request merged**